doc/content/xapi-guard/_index.md

```diff
@@ -26,3 +28,39 @@ Xapi-guard forwards calls from domains to xapi to persist UEFI variables, and up
 To do this, it listens to 1 socket per service (varstored, or swtpm) per domain.
 To create these sockets before the domains are running, it listens to a message-switch socket.
 This socket listens to calls from xenopsd, which orchestrates the domain creation.
+
+To protect the domains from xapi being unavailable transiently, xapi-guard provides an on-disk cache for vTPM writes.
+This cache stores the requests temporarily until xapi can be contacted again.
+This situation usually happens when xapi is being restarted as part of an update.
+
+The cache module consists of two Lwt threads, one that writes to disk, and another one that reads from disk.
+Both try to send the requests to xapi, depending on the state, to avoid using the cache if possible.
+The threads communicate through a bounded queue, this is done to limit the amount of memory used.
```

Comment:
...and the queue itself is essentially a performance optimisation, where the writer informs the reader precisely which are the names of the cache files, such that the reader does not need to list the cache directory. And a full queue does not mean data loss, just a loss of performance; vTPM writes are still cached.

Reply:
> When the queue has been filled the writer stops adding elements to the queue, and the reader will try to flush the queue, and after it will try to flush the cache. 

when the queue is full, the writer stops talking to the reader? which means that the reader needs to list the buffer directory to find out which files are buffered? 

> After this happens both threads will transition to cache bypass operation.

bypass the buffer and write to xapi db directly? But isn't this supposed to happen only when xapi is up again?

Reply:
> when the queue is full, the writer stops talking to the reader? which means that the reader needs to list the buffer directory to find out which files are buffered?

Yes, the reader loops through all directories and files and tries to push them if they match a valid key (vm uuid, timestamp, inner path) for the contents

> bypass the buffer and write to xapi db directly? But isn't this supposed to happen only when xapi is up again?
Yes, the transition can only happen when xapi can be written to. I'm not sure what's the contradiction in the statement, is it missing context?